tiledbcontents/async_tools.py

```diff
@@ -0,0 +1,119 @@
+"""Stuff used to run TileDB core code asynchronously in another process.
+
+Also includes some miscellaneous utilities.
+"""
+
+
+import asyncio
+import functools
+import os
+import pathlib
+import sys
+import tempfile
+from asyncio import subprocess
+from typing import Callable, Optional, TypeVar
+
+import cloudpickle
+from typing_extensions import ParamSpec, Self
+
+_P = ParamSpec("_P")
+_R = TypeVar("_R")
+
+
+async def call(__fn: Callable[_P, _R], *args: _P.args, **kwargs: _P.kwargs) -> _R:
+    """Calls ``__fn(*args, **kwargs)`` on an executor as to not block.
+
+    If you are making a call which eventually uses the core TileDB library,
+    **do not use this function**! Instead, use ``call_external``, to ensure
+    that your operations are fork-safe.
+    """
+    loop = asyncio.get_event_loop()
+    return await loop.run_in_executor(None, functools.partial(__fn, *args, **kwargs))
+
+
+_stub_lock = asyncio.Lock()
+"""Lock guarding the creation of the lazy stub."""
+_lazy_stub: Optional["_CallClient"] = None
+"""A lazily-created stub used to make external calls.
+
+Will be initialized the first time ``call_external`` is run.
+"""
+
+
+async def call_external(
+    __fn: Callable[_P, _R], *args: _P.args, **kwargs: _P.kwargs
+) -> _R:
+    """Calls ``__fn(*args, **kwargs)`` in an external process without blocking.
+
+    All calls which go to TileDB Core libraries need to be done in an external
+    process since TileDB Core is not fork-safe, and JupyterLab uses forking
+    to launch terminals.
+
+    All calls sent to this function are handled in separate threads within the
+    same external process.
+    """
+    global _lazy_stub
+    async with _stub_lock:
+        if _lazy_stub is None:
+            _lazy_stub = await _CallClient.create()
+    return await _lazy_stub.run(__fn, *args, **kwargs)
+
+
+class _CallClient:
+    """Stub to run callables in an external process."""
+
+    def __init__(self, proc: subprocess.Process, addr: pathlib.Path) -> None:
+        """Basic initializer. Use ``.create()`` to build a usable instance."""
+        self._proc = proc
+        """The async subprocess that is handling calls."""
```

Comment:
nit: doc above the definition?

Reply:
This is intentional; it’s the de facto standard for “fake docstrings” on things that can’t themselves have docstrings:

![The _proc variable with a tooltip containing “The async subprocess that is handling calls.”](https://github.com/TileDB-Inc/TileDB-Cloud-Jupyter-Contents/assets/6622003/71664a0c-56ef-4186-9456-be35fea58f6e)